Add a tutorial on porting test groups from legacy (#128)
MPAS-Dev/compass: docs/tutorials/dev_porting_legacy.rst

at the code and copying it as needed, you can simply browse the branch directly
on GitHub: https://github.com/MPAS-Dev/compass/tree/legacy/ocean/gotm/2.5km/default.
If you prefer, you can use either of the approaches described in
:ref:`dev_compass_repo` or :ref:`dev_compass_repo_advanced` to clone the repo
and check out the ``legacy`` branch.

Next, you will need to create a new branch from ``master`` for developing the
new test case.  For this purpose, we will stick with the simpler approach in
:ref:`dev_compass_repo` here, but feel free to use the ``worktree`` approach
instead if you are comfortable with it.

.. code-block:: bash

    git clone git@github.com:MPAS-Dev/compass.git add_gotm
    cd add_gotm

Now, you will need to create a conda environment for developing compass, as
described in :ref:`dev_conda_env`.  We will assume a simple situation where
you are working on a "supported" machine and using the default compilers and
MPI libraries, but consult the documentation to make an environment to suit
your needs.

.. code-block:: bash

  # this one will take a while the first time
  ./conda/create_compass_env.py --conda $HOME/miniconda3

> If all goes well, you will have a file named ``test_compass_1.0.0*.sh``, where

Author: ```suggestion
If all goes well, you will have a file named ``load_dev_compass_1.0.0*.sh``, where
```

                <argument flag="-o">grid.nc</argument>
            </step>
            <step executable="MpasCellCuller.x">
                <argument flag="">grid.nc</argument>
                <argument flag="">culled_mesh.nc</argument>
            </step>
            <step executable="MpasMeshConverter.x">
                <argument flag="">culled_mesh.nc</argument>
                <argument flag="">mesh.nc</argument>
            </step>
            <model_run procs="1" threads="1" namelist="namelist.ocean" streams="streams.ocean"/>
        </run_script>

    </config>

The XML files for the other steps look similar.  We will go through these files
in detail later in the tutorial.

The example test case also has a namelist file used by GOTM (``gotmturb.nml``)
and a python script for plotting the results compared to analytic solutions
(``plot_profile.py``).

Making a new test group
-----------------------

Okay, with those details as a reference point from legacy COMPASS, let's jump
into developing the new test group in ``compass``.  Use any method you like
for editing code.  If you haven't settled on a method and are working on your

> own laptop or desktop, you may want to try and integrated development

Author: ```suggestion
own laptop or desktop, you may want to try an integrated development
```

environment (`PyCharm <https://www.jetbrains.com/pycharm/>`_ is a really nice
one).  They have features to make sure your code adheres to the style required
for compass (see :ref:`dev_style`).  ``vim`` or a similar tool will work fine
on supercomputers.

In ``compass``, the ``gotm`` test group will be a new python package.  We will
make a new ``gotm`` directory in ``compass/ocean/tests``.  In that directory,
we will make a new, initially empty file ``__init__.py``.  Now, ``gotm`` is a
new package in ``compass`` that could be imported as

.. code-block:: python

    from compass.ocean.tests import gotm

Next, let's make a new class for the ``gotm`` test group in ``__init__.py``:

.. code-block:: python

    from compass.testgroup import TestGroup

    class Gotm(TestGroup):
        """

>         A test group for baroclinic channel test cases

Author: ```suggestion
        A test group for General Ocean Turbulence Model (GOTM) test cases
```

    <immutable_stream name="input_init"
                      filename_template="mesh.nc"/>

    <stream name="output_init"
            type="output"
            output_interval="0000_00:00:01"
            clobber_mode="truncate"
            filename_template="ocean.nc">

        <stream name="input_init"/>
        <var name="layerThickness"/>
        <var name="restingThickness"/>
        <var name="refBottomDepth"/>
        <var name="bottomDepth"/>
        <var name="maxLevelCell"/>
        <var name="vertCoordMovementWeights"/>
        <var name="edgeMask"/>
    </stream>

    </streams>

As in legacy COMPASS, streams that are already defined like ``input_init``
will use the default attributes defined by the MPAS component unless they are
explicitly replaced in the streams file.  On setting up the test case, the
stream in the ``streams.ocean`` file becomes:

.. code-block:: xml

    <immutable_stream name="input_init"
                      type="input"
                      filename_template="mesh.nc"
                      input_interval="initial_only"/>

Defining config options
~~~~~~~~~~~~~~~~~~~~~~~

The remainder of the ``init`` step will consist of defining the ``run()``

> method that does the refal work of the step.  To make it easier for users to

Reviewer: refal typo

Author: ```suggestion
method that does the real work of the step.  To make it easier for users to
```

modify the test case a little bit to suit their needs, we may want to include
parameters in the config file for the test case.  To do this, we can make a
config file with the test group's package, the test case's package, or both.
In our example, we will just add a config file ``defaults.cfg`` to the
``defaults`` test case:

.. code-block:: cfg

    # config options for General Ocean Turbulence Model (GOTM) test cases
    [gotm]

    # the number of grid cells in x and y
    nx = 4
    ny = 4

    # the size of grid cells (m)
    dc = 2500.0

    # the number of vertical levels
    vert_levels = 250

    # the depth of the sea floor (m)
    bottom_depth = 15.0

By default, the domain is 4 x 4 horizontal cells, each 2.5 km in size.  The
ocean is 15 m deep, divided over 250 uniformly spaced levels.  A user could
change any of these parameters before running the ``init`` step to modify
the initial condition (and therefore the rest of the test case).

Since the config file has the same name (``default``) as the test case, it will

> be included automatically when the config file produced when test case gets

Reviewer: ```suggestion
be included automatically when the config file is produced when the test case gets
```

        <var name="velocityMeridional"/>
        <var name="vertViscTopOfCell"/>
        <var name="xtime"/>
        <var name="normalVelocity"/>
        <var name="layerThickness"/>
    </stream>

    </streams>

The run script from the legacy test case:

.. code-block:: xml

    <run_script name="run.py">
        <model_run procs="1" threads="1" namelist="namelist.ocean" streams="streams.ocean"/>
    </run_script>

becomes:

.. code-block:: python

    from compass.model import run_model

    ...

    def run(self):
        """
        Run this step of the test case
        """
        run_model(self)

Adding the analysis step
------------------------

The legacy ``analysis`` step is defined like this:

.. code-block:: xml

    <config case="analysis">
        <add_link source="../forward/output.nc" dest="output.nc"/>
        <add_link source_path="script_test_dir" source="plot_profile.py" dest="plot_profile.py"/>

        <run_script name="run.py">
            <step executable="./plot_profile.py">
            </step>
        </run_script>
    </config>

Symlinks are created to a plotting script and the output from the ``forward``
step, and then the plot script is run.

> An identical approach could be used in ``compass`` but it is not the perferred

Reviewer: ```suggestion
An identical approach could be used in ``compass`` but it is not the preferred
```

test-case run (otherwise the data might not be available). The call to
:py:func:`from compass.validate.compare_variables()` will check whether
variables ``layerThickness`` and ``normalVelocity`` are exactly the same in
this run as they were in a previous run if a baseline run was provided when the
test case got set up (see :ref:`test_suites`).

Set up and run
--------------

You're all set!  You should be able to see your new test case when you run
``compass list``, set it up by running ``compass setup``, and run it by running
``compass run`` within the work directory.  See :ref:`dev_command_line` for
more on that process.

Documentation
-------------

Make sure to add some documentation of your new test group.  You need to add
all of the functions, classes and methods to the API documentation in
``docs/developers_guide/<core>/api.rst``, following the examples for other
test groups.  You also need to add a file to both the user's guide and the
developer's guide describing the test group and its test cases and steps.

For the user's guide, create a file
``docs/users_guide/<core>/test_groups/<test_group>.rst``.  In that file, you

> should describe what the test group and its test cases do in a way that would

Reviewer: ```suggestion
should describe what the test group and what its test cases do in a way that would
```

               out_name='namelist.ocean')

        run_model(self)

First, we make a doubly periodic mesh.  Rather than hard-coding the mesh size,
we get the relevant config options.  Legacy COMPASS used the command-line
tool ``planar_hex``, but ``compass`` will typically use the
:py:func:`mpas_tools.planar_hex.make_planar_hex_mesh()` funcion instead to
avoid the complexity of ``subprocess`` calls and unnecessary file I/O.  The
result is an :py:class:`xarray.Dataset` containing the mesh.

Second, we make sure any land cells are culled by calling the cell culler.
In legacy COMPASS, this is done with the command-line tool ``MpasCellCuller.x``
but in ``compass``, the same can be achieved with
:py:func:`mpas_tools.mesh.conversion.cull()` (which is a wrapper around a
``subprocess``` call to ``MpasCellCuller.x``).  ``cull()`` has
:py:class:`xarray.Dataset` objects as its input an return value, and also
takes a "logger" where it can write output (sometimes a log file and sometimes
directly to the terminal via stdout).  You should always pass ``self.logger``
so ``compass`` can figure out whether a file or stdout is the right place for
output to go.

.. note::

>     In this particular test case, the there are no land cells defined and

Reviewer 2: ```suggestion
    In this particular test case, there are no land cells defined and
```

        if 'forward' in self.steps_to_run:
            compare_variables(test_case=self,
                              variables=['layerThickness', 'normalVelocity'],
                              filename1='forward/output.nc')

We will only perform valiation if the ``forward`` step was part of the current
test-case run (otherwise the data might not be available). The call to
:py:func:`from compass.validate.compare_variables()` will check whether
variables ``layerThickness`` and ``normalVelocity`` are exactly the same in
this run as they were in a previous run if a baseline run was provided when the
test case got set up (see :ref:`test_suites`).

Set up and run
--------------

You're all set!  You should be able to see your new test case when you run
``compass list``, set it up by running ``compass setup``, and run it by running
``compass run`` within the work directory.  See :ref:`dev_command_line` for
more on that process.

Documentation
-------------

Make sure to add some documentation of your new test group.  You need to add
all of the functions, classes and methods to the API documentation in
``docs/developers_guide/<core>/api.rst``, following the exmaples for other

> test groups.  You also need to add a file to both the user's guid and the

Reviewer 2: ```suggestion
test groups.  You also need to add a file to both the user's guide and the
```
